system-addon/content-src/components/TopSites/TopSite.jsx

```diff
@@ -312,67 +312,61 @@ export class _TopSiteList extends React.PureComponent {
         if (index !== this.state.draggedIndex) {
           this.props.dispatch(ac.SendToMain({
             type: at.TOP_SITES_INSERT,
-            data: {site: {url: this.state.draggedSite.url, label: this.state.draggedTitle}, index}
+            data: {site: {url: this.state.draggedSite.url, label: this.state.draggedTitle}, index, draggedFromIndex: this.state.draggedIndex}
           }));
           this.userEvent("DROP", index);
         }
         break;
     }
   }
   _getTopSites() {
-    return this.props.TopSites.rows.slice(0, this.props.TopSitesCount);
+    let topSites = this.props.TopSites.rows.slice(0, this.props.TopSitesCount);
+    if (topSites.length < this.props.TopSitesCount) {
+      topSites[this.props.TopSitesCount - 1] = null;
+    }
+    return topSites;
   }
 
   /**
    * Make a preview of the topsites that will be the result of dropping the currently
    * dragged site at the specified index.
    */
   _makeTopSitesPreview(index) {
-    const preview = this._getTopSites();
-    this._fillOrLeaveHole(preview, this.state.draggedIndex);
-    this._insertSite(preview, Object.assign({}, this.state.draggedSite, {isPinned: true}), index);
-    return preview;
-  }
+    const topSites = this._getTopSites();
+    topSites[this.state.draggedIndex] = null;
+    const pinnedOnly = topSites.map(site => ((site && site.isPinned) ? site : null));
+    const unpinned = topSites.filter(site => site && !site.isPinned);
+    const siteToInsert = Object.assign({}, this.state.draggedSite, {isPinned: true});
+    if (!pinnedOnly[index]) {
+      pinnedOnly[index] = siteToInsert;
+    } else {
+      // Find the hole to shift the pinned site(s) towards. We shift towards the
+      // hole left by the site being dragged.
+      let holeIndex = index;
+      const indexStep = index > this.state.draggedIndex ? -1 : 1;
+      while (pinnedOnly[holeIndex]) {
+        holeIndex += indexStep;
```

Comment:
There was some shared `common/Reducers.jsm` pinning code before that we can try to clean up in a followup.